src/libstd/primitive_docs.rs

```diff
@@ -67,6 +67,110 @@
+///     Some(num) => num,
+///     None => break,
+/// }
+/// ```
+///
+/// Both match arms must produce values of type `u32`, but since `break` never produces a value at
+/// all we know it can never produce a value which isn't a `u32`. This illustrates another
+/// behaviour of the `!` type - expressions with type `!` will coerce into any other type.
+///
+/// [`exit`]: process/fn.exit.html
+///
+/// # `!` and generics
+///
+/// The main place you'll see `!` used explicitly is in generic code. Consider the [`FromStr`]
+/// trait:
+///
+/// ```
+/// trait FromStr {
+///     type Error;
+///     fn from_str(s: &str) -> Result<Self, Self::Error>;
+/// }
+/// ```
+///
+/// When implementing this trait for `String` we need to pick a type for `Error`. And since
+/// converting a string into a string will never result in an error, the appropriate type is `!`.
+/// If we have to call `String::from_str` for some reason, the result will be a
+/// `Result<String, !>`, which we can unpack like this:
+///
+/// ```
+/// let Ok(s) = String::from_str("hello");
+/// ```
+///
+/// Since the `Err` variant contains a `!`, it can never occur. So we can exhaustively match on
+/// `Result<T, !>` by just taking the `Ok` variant. This illustrates another behaviour of `!` - it
+/// can be used to "delete" certain enum variants from generic types like `Result`.
+///
+/// [`FromStr`]: str/trait.FromStr.html
+///
+/// # `!` and traits
+///
+/// When writing your own traits, `!` should have an `impl` whenever there is an obvious `impl`
+/// which doesn't `panic!`. As is turns out, most traits can have an `impl` for `!`. Take [`Debug`]
+/// for example:
+///
+/// ```
+/// impl Debug for ! {
+///     fn fmt(&self, formatter: &mut fmt::Formatter) -> fmt::Result {
+///         *self
+///     }
+/// }
+/// ```
+///
+/// Once again we're using `!`'s ability to coerce into any other type, in this case `fmt::Result`.
+/// Since this method takes a `&!` as an argument we know that it can never be called (because
+/// there is no value of type `!` for it to be called with). Writing `*self` essentially tells the
+/// compiler "We know that this code can never be run, so just treat the entire function body has
+/// having type `fmt::Result`". This pattern can be used a lot when implementing traits for `!`.
+/// Generally, any trait which only has methods which take a `self` parameter should have such as
+/// impl.
+///
+/// On the other hand, one trait which would not be appropriate to implement is [`Default`]:
+///
+/// ```
+/// trait Default {
+///     fn default() -> Self;
+/// }
+/// ```
+///
+/// Since `!` has no values, it has no default value either. It's true that we could write an
+/// `impl` for this which simply panics, but the same is true for any type (we could `impl
+/// Default` for (eg.) `File` by just making `default()` panic.)
+///
+/// [`Debug`]: fmt/trait.Debug.html
+/// [`Default`]: default/trait.Default.html
+///
+#[stable(feature = "rust1", since = "1.23.0")]
```

Comment:
Should this be attached to `rust1`, or would it be attached to `never_type`? I would assume the latter.

Reply:
That's also the reason why tidy errors:
```
[00:03:19] tidy error: /checkout/src/libstd/primitive_docs.rs:171: mismatches to previous in: ["since"]
[00:03:19] tidy error: /checkout/src/libstd/primitive_docs.rs:247: mismatches to previous in: ["since"]
```

because rust1 has been defined elsewhere already (probably with a since field of "1.0.0") and the since field that @canndrew specified "1.23.0" mismatches.

Reply:
I changed it to `#[unstable(feature = "never_type_impls", issue = "35121")]` since that's what's used in other places in libstd/libcore.